### PR TITLE
Skip image loading if the image is in cache

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 1]
```diff
@@ -149,6 +150,15 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
       const auto file_name = in.GetSourceInfo();
 
       ImageInfo info;
+
+      const ImageCache::ImageShape kInvalidShape{};
+      auto cached_shape = CacheImageShape(file_name);
+      if (cached_shape != kInvalidShape) {
```
Comment (Kh4L) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]:
```suggestion
      if (cached_shape.empty()) {
```

Comment (jantonguirao) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: will do

**Comment by szalpal** [comment]
How do we verify, that for the reader implementations, skipping frames works properly? (i.e. frames that should be skipped are skipped, and frames that shouldn't, aren't?)

File: dali/pipeline/operators/reader/loader/recordio_loader.h [hunk 2]
```diff
@@ -94,6 +94,7 @@ class RecordIOLoader : public IndexedFileLoader {
       tensor.set_type(TypeInfo::Create<uint8_t>());
       tensor.Resize({1});
       tensor.SetSkipSample(true);
+      current_file_->Seek(seek_pos + size);
```
Comment (JanuszL) [inline dali/pipeline/operators/reader/loader/recordio_loader.h]: What if you seek past the end of file? Please check L125 - it seems that image can be splitted between files, or even be in the next file but you won't know that until you try to read file and get n_read  == 0.

Comment (jantonguirao) [reply dali/pipeline/operators/reader/loader/recordio_loader.h]: done

**Comment by szalpal** [comment]
> @szalpal I am not sure what you mean. Are you asking how do we test this? Or how do we enforce it in the next operators?
> 
> In general, this is a performance optimization. The output of the reader is to be ignored if the image is in decoder cache, therefore we can skip reading the file altogether and save some time

To clarify: How can we make sure, that frames that are actually in cache are omitted and frames that aren't in cache are kept?

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 3]
```diff
@@ -149,6 +150,14 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
       const auto file_name = in.GetSourceInfo();
 
       ImageInfo info;
+
+      auto cached_shape = CacheImageShape(file_name);
+      if (volume(cached_shape) > 0) {
+        output_shape_[i] = Dims({cached_shape[0], cached_shape[1], cached_shape[2]});
+        output_info_[i] = info;
```
Comment (Kh4L) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Do we need this reset of `output_info`? I think we can move L152 after this branch

Comment (jantonguirao) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: will do

**Comment by jantonguirao** [comment]
Build 691033

**Pull request merged**